software/contrib/egressus_melodiam.py

```diff
@@ -0,0 +1,1091 @@
+        return self.t
+
+    """Entry point - main loop"""
+
+    def main(self):
+        while True:
+            self.updateScreen()
+            self.getK1Value()
+            self.getOutputDivision()
+
+            # Cycle through outputs, process when needed
+            for idx in range(len(cvs)):
+                if (
+                    ticks_diff(ticks_ms(), self.lastSlewVoltageOutputTime[idx])
+                    >= self.msBetweenSamples[idx]
+                    and self.running
+                ):
+
+                    try:
+
+                        # Do we have a sample in the buffer?
+                        if self.slewBufferPosition[idx] < self.slewBufferSampleNum[idx]:
+                            # Yes, we have a sample, output voltage to match the sample, reset underrun counter and advance position in buffer
+
+                            # If a square interpolation mode (0) a precalculated value is used
+                            # as the generator function for square waves was really buggy
+                            if self.outputSlewModes[idx] == 0:
+                                v = self.squareOutputs[idx]
+                            else:
+                                v = next(self.slewGeneratorObjects[idx])
+                            cvs[idx].voltage(v)
+                            self.previousOutputVoltage[idx] = v
+                            self.bufferUnderrunCounter[idx] = 0
+                        else:
+                            # We do not have a sample - buffer under run
+                            # Output the previous voltage to keep things as smooth as possible
+                            cvs[idx].voltage(self.previousOutputVoltage[idx])
+                            self.bufferUnderrunCounter[idx] += 1
+                            if DEBUG_MODE == 2:
+                                diff = self.slewBufferPosition[idx] - self.slewBufferSampleNum[idx]
+                                # Print an error if we are 2 or more samples out
+                                if diff >= 2:
+                                    print(
+                                        f"[{self.clockStep}][{idx}] Buffer underrun [{self.bufferUnderrunCounter[idx]}] ({diff}). Using previous voltage: {self.previousOutputVoltage[idx]}"
+                                    )
+                                    print(
+                                        f"[{self.clockStep}][{idx}] Current sample num: {self.slewBufferPosition[idx]} Buffer size: {self.slewBufferSampleNum[idx]}"
+                                    )
+
+                        # Advance the position in the sample/slew buffer
+                        self.slewBufferPosition[idx] += 1
+
+                    except StopIteration:
+                        # we shouldn't ever get here...
+                        if DEBUG_MODE == 2:
+                            print(f"[{self.clockStep}][{idx}] ERROR: StopIteration")
+                        continue

```

Comment:
[discussion]
If we should never get here, can this code be safely removed?

Reply:
fair point; removed in favour of keeping the catch all at the end